File: move_utils.py
ATAQUES_DOIS_TURNOS = {
    "Solar Beam",
    "Hyper Beam",
    "Giga Impact",
    "Bounce",
    "Dive",
    "Dig",
    "Frenzy Plant",
    "Solar Blade",
    "Future Sight",
    "Rock Wrecker",
    "Meteor Beam",
    "Meteor Assault",
    "Fly",
    "Double Shock",
    "Phantom Force",
    "Gigaton Hammer",
}
MENOS_UM_STATS = {"Superpower"}
MENOS_DOIS_STATS = {"Overheat", "Leaf Storm"}
DANO_EM_SI_MESMO = {
    "Flare Blitz",
    "Double-Edge",
    "Axe Kick",
    "Brave Bird",
    "Wild Charge",
    "Wood Hammer",
    "Supercell Slam",
    "Light of Ruin",
}
BANIDOS = {
    "Focus Punch",
    "First Impression",
    "Fake Out",
    "Explosion",
    "Self-Destruct",
    "Steel Roller",
    "Dream Eater",
    "Misty Explosion",
}  # IA não consegue garantir que não vai tomar dano no turno de preparo
TRAVAMENTO_E_CONFUSAO = {"Outrage", "Thrash", "Petal Dance", "Raging Fury"}

# moves com hits variaveis: cada hit tem chance acc de continuar, de min a max hits
# effective_power sera ajustado pelo valor esperado de hits
MULTI_HIT = {
    "Population Bomb": (1, 10),
    "Bullet Seed": (2, 5),
    "Rock Blast": (2, 5),
    "Icicle Spear": (2, 5),
    "Pin Missile": (2, 5),
    "Bone Rush": (2, 5),
    "Fury Attack": (2, 5),
    "Fury Swipes": (2, 5),
    "Comet Punch": (2, 5),
    "Tail Slap": (2, 5),
    "Water Shuriken": (2, 5),
    "Bonemerang": (2, 2),
    "Gear Grind": (2, 2),
    "Twineedle": (2, 2),
}


def _hits_esperados(min_hits: int, max_hits: int, acc: float) -> float:
    """Valor esperado de hits para um move multi-hit com acc por hit."""
    p = acc / 100.0
    ev = 0.0
    prob_chegou = 1.0
    for k in range(1, max_hits + 1):
        if k < max_hits:
            prob_exato = prob_chegou * p * (1 - p)
        else:
            prob_exato = prob_chegou * p
        if k >= min_hits:
            ev += k * prob_exato
        prob_chegou *= p
    return ev


def normalizar_dano_por_turno(movepool):
    """Ajusta effective_power para refletir o custo real por turno e remove moves banidos."""
    resultado = []
    for move in movepool:
        if move.name in BANIDOS or move.power <= 0:
            continue
        if move.name in ATAQUES_DOIS_TURNOS:
            move.effective_power = move.power / 2.0
            move.turns = 2
            if "2T" not in move.tags:
                move.tags.append("2T")
        elif move.name in MENOS_UM_STATS:
            move.effective_power = move.power * 0.835
            if "-1ST" not in move.tags:
                move.tags.append("-1ST")
        elif move.name in MENOS_DOIS_STATS:
            move.effective_power = move.power * 0.75
            if "-2ST" not in move.tags:
                move.tags.append("-2ST")
        elif move.name in TRAVAMENTO_E_CONFUSAO:
            move.effective_power = move.power * 0.75
            if "LOCK" not in move.tags:
                move.tags.append("LOCK")
        elif move.name in DANO_EM_SI_MESMO:
            move.effective_power = move.power * 0.66
            if "SD" not in move.tags:
                move.tags.append("SD")
        elif move.name in MULTI_HIT:
            min_h, max_h = MULTI_HIT[move.name]
            hits = _hits_esperados(min_h, max_h, move.acc)
            move.effective_power = move.power * hits
            tag = f"x{min_h}-{max_h}" if min_h != max_h else f"x{min_h}"
            if tag not in move.tags:
                move.tags.append(tag)
        resultado.append(move)
    return resultado


def podar_estritamente_dominados(movepool):
    """Remove moves obsoletos pelo valor esperado (effective_power * acc) por tipo+categoria."""
    from collections import defaultdict

    def ev(move):
        return move.effective_power * (move.acc / 100.0) ** 2

    # Para cada (type, category), guarda o melhor EV e o nome do vencedor (desempate)
    melhor_por_grupo: dict = defaultdict(lambda: (-1.0, ""))
    for m in movepool:
        chave = (m.type, m.category)
        ev_m = ev(m)
        ev_atual, nome_atual = melhor_por_grupo[chave]
        if ev_m > ev_atual or (ev_m == ev_atual and m.name < nome_atual):
            melhor_por_grupo[chave] = (ev_m, m.name)

    return [m for m in movepool if melhor_por_grupo[(m.type, m.category)][1] == m.name]

File: typescript/app/src/engine/moveUtils.ts
import { Move } from './move'

const ATAQUES_DOIS_TURNOS = new Set([
  'Solar Beam', 'Hyper Beam', 'Giga Impact', 'Bounce', 'Dive', 'Dig',
  'Frenzy Plant', 'Solar Blade', 'Future Sight', 'Rock Wrecker',
  'Meteor Beam', 'Meteor Assault', 'Fly', 'Double Shock', 'Phantom Force', 'Sky Attack'
])
const MENOS_UM_STATS = new Set(['Superpower'])
const MENOS_DOIS_STATS = new Set(['Overheat', 'Leaf Storm'])
const DANO_EM_SI_MESMO = new Set([
  'Flare Blitz', 'Double-Edge', 'Axe Kick', 'Brave Bird',
  'Wild Charge', 'Wood Hammer', 'Supercell Slam',
])
const BANIDOS = new Set([
  'Focus Punch', 'First Impression', 'Fake Out', 'Explosion',
  'Self-Destruct', 'Steel Roller', 'Dream Eater', 'Misty Explosion','Foul Play', 'Psystrike', "Last Resort", "Belch"
])
const TRAVAMENTO_E_CONFUSAO = new Set(['Outrage', 'Thrash', 'Petal Dance', 'Raging Fury'])

// Tipagem estruturada para separar a mecânica estatística
type MultiHitMechanic = 'classic' | 'continuous' | 'escaling' | 'fixed';

interface MultiHitConfig {
  type: MultiHitMechanic;
  min: number;
  max: number;
}

const MULTI_HIT: Record<string, MultiHitConfig> = {
  // Accuracy rolada a cada hit
  'Population Bomb': { type: 'continuous', min: 1, max: 10 },
  'Triple Dive':     { type: 'continuous', min: 1, max: 3 },
  
  // Accuracy rolada a cada hit + O Poder multiplica a cada hit
  'Triple Axel':     { type: 'escaling', min: 1, max: 3 },

  // Accuracy rolada 1 vez no início. RNG interno define a quantidade de hits (3.1 média)
  'Bullet Seed':     { type: 'classic', min: 2, max: 5 },
  'Rock Blast':      { type: 'classic', min: 2, max: 5 },
  'Icicle Spear':    { type: 'classic', min: 2, max: 5 },
  'Pin Missile':     { type: 'classic', min: 2, max: 5 },
  'Bone Rush':       { type: 'classic', min: 2, max: 5 },
  'Fury Attack':     { type: 'classic', min: 2, max: 5 },
  'Fury Swipes':     { type: 'classic', min: 2, max: 5 },
  'Comet Punch':     { type: 'classic', min: 2, max: 5 },
  'Tail Slap':       { type: 'classic', min: 2, max: 5 },
  'Water Shuriken':  { type: 'classic', min: 2, max: 5 },

  // Hits cravados (2 hits sempre)
  'Bonemerang':      { type: 'fixed', min: 2, max: 2 },
  'Gear Grind':      { type: 'fixed', min: 2, max: 2 },
  'Twineedle':       { type: 'fixed', min: 2, max: 2 },
  'Dual Chop':       { type: 'fixed', min: 2, max: 2 },
  'Twin Beam':        { type: 'fixed', min: 2, max: 2 },
}

// Progressão geométrica: Retorna a média de hits antes de errar
function hitsEsperadosContinuos(minHits: number, maxHits: number, acc: number): number {
  const p = acc / 100.0
  let ev = 0.0
  let probChegou = 1.0
  for (let k = 1; k <= maxHits; k++) {
    const probExato = k < maxHits ? probChegou * p * (1 - p) : probChegou * p
    if (k >= minHits) ev += k * probExato
    probChegou *= p
  }
  return ev
}

export function normalizarDanoPorTurno(movepool: Move[], banirRecoil: boolean = false, banirLock: boolean = false): Move[] {
  const resultado: Move[] = []

  for (const move of movepool) {
    if (BANIDOS.has(move.name) || move.power <= 0) continue
    if (banirRecoil && DANO_EM_SI_MESMO.has(move.name)) continue
    if (banirLock && TRAVAMENTO_E_CONFUSAO.has(move.name)) continue

    if (ATAQUES_DOIS_TURNOS.has(move.name)) {
      move.effectivePower = move.power / 2.0
      move.turns = 2
      if (!move.tags.includes('2T')) move.tags.push('2T')
    } else if (MENOS_UM_STATS.has(move.name)) {
      move.effectivePower = move.power * 0.835
      if (!move.tags.includes('-1ST')) move.tags.push('-1ST')
    } else if (MENOS_DOIS_STATS.has(move.name)) {
      move.effectivePower = move.power * 0.75
      if (!move.tags.includes('-2ST')) move.tags.push('-2ST')
    } else if (TRAVAMENTO_E_CONFUSAO.has(move.name)) {
      move.effectivePower = move.power * 0.75
      if (!move.tags.includes('LOCK')) move.tags.push('LOCK')
    } else if (DANO_EM_SI_MESMO.has(move.name)) {
      move.effectivePower = move.power * 0.66
      if (!move.tags.includes('SD')) move.tags.push('SD')
    } else if (move.name in MULTI_HIT) {
      const config = MULTI_HIT[move.name]
      let finalEffectivePower = move.power

      if (config.type === 'classic') {
        // RNG da Gen 5+: A média cravada é 3.1 hits.
        const evHits = config.max === 5 ? 3.1 : config.max
        finalEffectivePower = move.power * evHits
      } else if (config.type === 'fixed') {
        finalEffectivePower = move.power * config.max
      } else if (config.type === 'continuous') {
        // Ex: Population Bomb (Média ~6.5 hits se acc 90)
        const evHits = hitsEsperadosContinuos(config.min, config.max, move.acc)
        finalEffectivePower = move.power * evHits
      } else if (config.type === 'escaling') {
        // Ex: Triple Axel. (Power 20 -> 40 -> 60) multiplicados pela probabilidade de cada hit conectar
        const p = move.acc / 100.0
        let evPower = 0
        let probChegou = 1.0
        for (let k = 1; k <= config.max; k++) {
          probChegou *= p // chance do hit K conectar
          evPower += (move.power * k) * probChegou 
        }
        finalEffectivePower = evPower
      }

      move.effectivePower = finalEffectivePower
      const tag = config.min !== config.max ? `x${config.min}-${config.max}` : `x${config.min}`
      if (!move.tags.includes(tag)) move.tags.push(tag)
    }

    resultado.push(move)
  }
  return resultado
}

export function podarEstritamenteDominados(movepool: Move[]): Move[] {
  const ev = (m: Move) => m.effectivePower * (m.acc / 100) ** 2

  const melhorPorGrupo = new Map<string, { ev: number; name: string }>()
  for (const m of movepool) {
    const chave = `${m.type}|${m.category}`
    const evM = ev(m)
    const atual = melhorPorGrupo.get(chave)
    if (!atual || evM > atual.ev || (evM === atual.ev && m.name < atual.name)) {
      melhorPorGrupo.set(chave, { ev: evM, name: m.name })
    }
  }

  return movepool.filter(m => melhorPorGrupo.get(`${m.type}|${m.category}`)?.name === m.name)
}